### ISSUE TITLE
Finalize the DriftStream API

### ISSUE THREAD

**Issue by hmgomes** [adaptive-machine-learning/CapyMOA#90]
The DriftStram API is still missing a few bells and whistles. 
The most relevant one is the DriftStream Range strategy. The implementation will be similar to the DriftStream Position strategy. 

More details on `04_drift_stream.ipynb`

**Comment by hmgomes** [comment]
This was put off by #179 , so it will be resumed now that the issue was closed

**Comment by hmgomes** [comment]
## Status check — what "Range" means, and is it blocked?

Went through `notebooks/04_drift_streams.ipynb` (§4.2 and §4.2.4, still showing the *under construction* image) and the current `stream/drift.py`. Summary: **not blocked, and it does not need major changes to the drift simulation API** — but there is a prerequisite worth fixing first.

### The proposed API, for the record

The notebook specifies Range as durations rather than absolute offsets:

```python
DriftStream([
    Concept(SEA(1), num_instances=1000),
    AbruptDrift(),
    Concept(SEA(2), num_instances=500),
    GradualDrift(num_instances=500),
    Concept(SEA(3), num_instances=500),
])
```

The `Concept()` wrapper exists because `Stream` (and therefore `SEA` and the other generators) has no `num_instances`/`max_instances` notion, and adding one would break MOA compatibility. The notebook is also explicit that the two styles **must not be mixed**: in Range mode the specification does not state where drifts land, so allowing a `position` alongside it would be error-prone.

### Implementation assessment

The existing builder walks the list and emits a nested MOA `ConceptDriftStream` CLI using `-p {drift.position} -w {drift.width}`. Range mode is a **front-end translation**: accumulate `num_instances` into absolute positions, then hand off to the CLI generation that already exists, unchanged. Concretely:

- add a `Concept` wrapper (stream + `num_instances`)
- `AbruptDrift.__init__` currently takes `position: int` as a **required** argument — must become optional
- `GradualDrift` needs `num_instances` as an alternative to `position`/`width` and `start`/`end`
- validation rejecting mixed styles
- a pre-pass converting durations to positions

No change to the MOA CLI layer. That is the good news.

### Prerequisite: `GradualDrift.__init__` is already fragile

It currently branches on `start is None and end is None` / `position is None and width is None`, and two call shapes fall through:

- `GradualDrift(position=100)` — takes the first branch with `width=None`, then `int(position - width / 2)` raises `TypeError`
- `GradualDrift(position=100, start=95)` — matches **neither** branch, so `self.width`/`self.position` are never assigned and it fails later with `AttributeError`

Range mode adds a **third** construction style to the same constructor. Adding it on top of the current branching will make this worse. Suggest restructuring the constructor (explicit "exactly one of position+width / start+end / num_instances" validation) as a small, separate PR **before** Range lands.

Related: `get_recurrent_concept_drift_stream_list` reads `drift_args["position"]` and scales it per recurrence, so `RecurrentConceptDriftStream` assumes absolute positions. Range mode needs a decision about whether it composes with recurrence at all.

### Open question

In `position` mode the final concept is deliberately unbounded — the user bounds the stream externally via `max_instances`. In Range mode every concept has a length, so the stream has an implied total. Should `DriftStream` expose that (e.g. a `length`), and what happens if the user then passes a larger `max_instances`? Worth deciding before implementation, since it changes the class contract.

## Relationship to the drift-evaluation paper and @vcerqueira's experiments

Checked `vcerqueira/experiments-drift_evaluation` (`src/streams/inject_drift.py`) against this API. **They are orthogonal mechanisms and should stay separate:**

| | `DriftStream` (this issue) | `DriftSimulator` (paper / experiments repo) |
|---|---|---|
| mechanism | concept **substitution** — switch between generators | instance **transformation** — `g(I)` applied to instances |
| data | synthetic generators | real-world datasets |
| drift kinds | whatever the generators differ in | class prior, label swap, feature permutation, feature filtering |
| API shape | declarative list of concepts + drifts | `fit(stream_size)` / `transform(instance)` |
| drift location | specified by the user | sampled per Monte Carlo trial within a region |

They share vocabulary — `(d_s, d_e)`, width `w`, abrupt as `w = 0` — but a transformation-based injector cannot be expressed as a list of concepts, so folding it into `DriftStream` would distort both.

**One concrete divergence worth deciding on.** The gradual transition is defined differently on each side:

- paper / `inject_drift.py`: **linear** ramp — `progress = (idx - onset) / width`, apply with probability `progress`
- MOA `ConceptDriftStream`: a **sigmoid**, per Bifet et al. (2011) §2.7.1, which is the reference already cited in our `Drift` docstring, with `alpha` as the grade of change

So `GradualDrift(width=500)` in CapyMOA and `width=500` in the experiments repo do not describe the same transition shape. I tried to confirm this empirically by measuring the class-mix curve across a `GradualDrift` window between two SEA functions, but the two concepts differ by only ~0.08 in class rate, so linear and sigmoid predictions fell within sampling noise — **inconclusive, and worth confirming against the MOA source** before we document any equivalence.

### Suggestion

Keep this issue scoped to the Range builder, and track the transformation-based injector separately — the paper's simulation framework is a genuinely useful thing to have in CapyMOA, but as `capymoa.drift.simulation` (or similar) rather than as part of `DriftStream`. Happy to open that issue if you want it tracked.

No PR opened for either.

**Comment by hmgomes** [comment]
Filed the constructor prerequisite as a private backlog item: `adaptive-machine-learning/backlog#100` — *GradualDrift constructor validation is incomplete*.

Confirmed there that **5 of 8** constructor shapes currently fail with an internal `TypeError`/`AttributeError` rather than a clear message, including `GradualDrift()` with no arguments — which is close to the shape Range mode needs. That lands first; this issue stays scoped to the Range builder.

**Comment by hmgomes** [comment]
```
## Relationship to the drift-evaluation paper and @vcerqueira's experiments

Checked `vcerqueira/experiments-drift_evaluation` (`src/streams/inject_drift.py`) against this API. **They are orthogonal mechanisms and should stay separate:**

| | `DriftStream` (this issue) | `DriftSimulator` (paper / experiments repo) |
|---|---|---|
| mechanism | concept **substitution** — switch between generators | instance **transformation** — `g(I)` applied to instances |
| data | synthetic generators | real-world datasets |
| drift kinds | whatever the generators differ in | class prior, label swap, feature permutation, feature filtering |
| API shape | declarative list of concepts + drifts | `fit(stream_size)` / `transform(instance)` |
| drift location | specified by the user | sampled per Monte Carlo trial within a region |

They share vocabulary — `(d_s, d_e)`, width `w`, abrupt as `w = 0` — but a transformation-based injector cannot be expressed as a list of concepts, so folding it into `DriftStream` would distort both.

```

Regarding this part, `GradualDrift` can incorporate some mechanism that let the user decide the function to be used to transition from one concept to the other. By default, we use the sigmoid, but nothing prevents us from using the linear transition, or even others (perhaps even accepting a lambda or similar, whichever is more API rich without causing confusion, supporting initially linear and sigmoid, should be ok, my "programmer" brain thinks a parameter to the function like "transition_function=["sigmoid", "linear", lambda]" so it could be different types. That is probably a terrible approach to the problem. This might be part of a separate issue, handled before this issue, however since it influences the `range` definition as well, this issue should be aware of it or at least the new one be aware of it

**Comment by hmgomes** [comment]
## On a configurable transition function

Dug into where the transition actually executes, because it decides whether this can be a parameter at all.

**It cannot, today.** `moa.streams.ConceptDriftStream` exposes only six options — there is no transition-function knob:

```
-s stream   -d driftstream   -a alpha (Angle alpha of change grade)
-p position -w width         -r randomSeed
```

and the mixing is compiled Java inside `nextInstance()`. Disassembling it:

```
-4.0 * (numberInstanceStream - position) / width
Math.exp(...)                    ->  1.0 / (1.0 + exp(...))
random.nextDouble()  ->  dcmpl / ifle  ->  inputStream : driftStream
```

That settles the open question from my earlier comment: it is the **logistic sigmoid** `p(t) = 1 / (1 + e^{-4(t-p)/w})`, drawn per instance as a Bernoulli trial. Confirmed from the bytecode rather than inferred — my two attempts to measure it empirically were inconclusive because the class-rate gap between concepts was too small.

So a `transition_function=` parameter added on the Python side would be **silently ignored** — MOA would keep doing the sigmoid regardless. That is worse than not having the parameter, which is why I would keep it out of #372 (which changes no behaviour at all) and out of any change that cannot honour it.

### Two routes, and they differ a lot

**(a) Java route** — extend/replace `ConceptDriftStream` in MOA, refresh `moa.jar`. Keeps the current CLI-composition model and the per-instance performance, but ties the feature to a MOA release cycle (`invoke refresh_moa`), and a Python lambda can never be expressed in a MOA CLI — so this route supports named functions only.

**(b) Python route** — move the Bernoulli draw into CapyMOA: read from both concepts and choose per instance in Python, instead of delegating gradual drift to `ConceptDriftStream`. This makes the transition a first-class Python concept, so a callable becomes possible, and it aligns with how the drift-evaluation paper and `vcerqueira/experiments-drift_evaluation` do it (linear ramp, `progress = (idx - onset) / width`). Costs: per-instance Python overhead on a hot path, and `DriftStream` stops being expressible as a single MOA CLI, which affects `get_moa_stream()` and the `_CLI` string that tests currently assert on.

### On the proposed API shape

`transition_function=["sigmoid", "linear", lambda]` is not a terrible approach — a `Literal[...] | Callable` union is a well-precedented pattern (e.g. scikit-learn's `metric=`). The shape is fine; **the blocker is execution location, not the signature.** If route (a) is taken, the callable arm cannot exist; if route (b) is taken, it comes for free.

If it does become a callable, worth pinning the contract early: `f(progress: float) -> float` mapping `[0, 1]` to a probability of drawing from the new concept, so `linear` is `lambda p: p` and the current behaviour is `lambda p: 1/(1+exp(-4*(p-0.5)))`. That keeps it testable and makes `"sigmoid"`/`"linear"` just named shorthands.

### Separate finding worth its own fix: `alpha` silently overrides `width`

While reading the bytecode: `alpha` is **not** used in the mixing at all. It is read once in `prepareForUseImpl`, where `alpha != 0` causes `width` to be **overwritten** with `1/tan(alpha)`. Our `Drift` docstring describes it only as *"The grade of change, defaults to 0.0"*, and `DriftStream` always emits `-a {drift.alpha}`.

So `GradualDrift(position=1000, width=500, alpha=0.1)` silently discards `width=500`. It is harmless at the default of `0.0`, but any user who sets `alpha` gets a drift width they did not ask for, with nothing saying so. Small, self-contained fix — happy to file it.

### Ordering

I would not gate the Range builder on this. They are different axes: Range changes how *positions are specified*, the transition function changes how *instances are mixed*. `num_instances` and `transition_function` are independent parameters on the same constructor, and #372 makes adding both safe.

**But** if route (b) is chosen, `DriftStream` stops composing a nested MOA CLI — and that CLI is exactly what Range's duration-to-position translation targets. So the *decision* between (a) and (b) should be made before Range is **implemented**, even if the transition work merges later. That is the real coupling, rather than the parameters themselves.

**Comment by hmgomes** [comment]
## Proposal: re-scope this issue around the composition model

Following the discussion above, there is a more fundamental constraint than the Range builder, and it explains several problems at once: **`DriftStream` delegates composition to MOA's `ConceptDriftStream`.**

The consequence I had not appreciated until testing it — Python-native streams cannot be concepts at all, because `ConceptDriftStream` takes MOA streams. `NumpyStream`, `CSVStream` and `TorchStream` are not `MOAStream` subclasses, and the builder loop skipped anything it did not recognise:

```python
DriftStream([SEA(1), AbruptDrift(position=250), NumpyStream(...)])
```
```
constructed WITHOUT error
declared drifts : 1                              <- reports a drift
CLI             :  -s (generators.SEAGenerator)  <- 2nd concept and drift gone
```

A stream that never drifts while advertising a drift. adaptive-machine-learning/CapyMOA#372 now makes that an error; making it *work* needs the composition to move into Python.

That single change also resolves the other threads on this issue:

- the **transition function** becomes reachable (it is currently compiled Java — see the bytecode in the comment above), so `"sigmoid" | "linear" | callable` is possible
- the **Range builder** loses its duration→position→CLI translation and becomes an accumulator
- the paper's **transformation-based injection** gets a natural home as a sibling of substitution-based drift
- **`alpha`**, which silently overrides `width` in MOA, stops being something we have to inherit

Written up with the evidence, the costs (reproducibility, performance, the `MOAStream` type contract, the two `_CLI` assertions from #362, `RecurrentConceptDriftStream`) and open questions as **`adaptive-machine-learning/backlog#101`** — *Compose DriftStream in Python instead of delegating to MOA ConceptDriftStream*.

Suggested framing: **backlog#101 becomes the parent, and this issue stays the Range builder deliverable under it.** The Range work does not need to wait for the whole redesign to merge, but the composition decision should be made before Range is implemented — otherwise Range gets built against a CLI translation layer that is about to be removed.

Order that seems sensible:

1. CapyMOA#372 — construction validation (open, CI green)
2. backlog#101 item 1 — Python composition, default behaviour preserved
3. transition function
4. **this issue** — Range builder
5. transformation-based injection

Happy to be argued out of the ordering; the part I would not change is deciding 2 before implementing 4.
